server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -54,134 +56,204 @@
+                final Manifest manifest = metaStateAndData.v1();
+                final MetaData metaData = metaStateAndData.v2();
+
                 // We finished global state validation and successfully checked all indices for backward compatibility
                 // and found no non-upgradable indices, which means the upgrade can continue.
                 // Now it's safe to overwrite global and index metadata.
-                if (metaData != upgradedMetaData) {
+                long globalStateGeneration = manifest.getGlobalGeneration();
+
+                if (globalStateGeneration != -1) {
+                    // If globalStateGeneration is non-negative, it means we should have some metadata on disk
+                    // Always re-write it even if upgrade plugins do not upgrade it, to be sure it's properly persisted on all data path
+
+                    List<Runnable> cleanupActions = new ArrayList<>();
+                    final MetaData upgradedMetaData = upgradeMetaData(metaData, metaDataIndexUpgradeService, metaDataUpgrader);
+
                     if (MetaData.isGlobalStateEquals(metaData, upgradedMetaData) == false) {
-                        metaStateService.writeGlobalState("upgrade", upgradedMetaData);
+                        globalStateGeneration = metaStateService.writeGlobalState("upgrade", upgradedMetaData);
+                    } else {
+                        globalStateGeneration = metaStateService.writeGlobalState("startup", metaData);
                     }
+                    final long currentGlobalStateGeneration = globalStateGeneration;
+                    cleanupActions.add(() -> metaStateService.cleanupGlobalState(currentGlobalStateGeneration));
+
+                    Map<Index, Long> indices = new HashMap<>();
                     for (IndexMetaData indexMetaData : upgradedMetaData) {
+                        long generation;
                         if (metaData.hasIndexMetaData(indexMetaData) == false) {
-                            metaStateService.writeIndex("upgrade", indexMetaData);
+                            generation = metaStateService.writeIndex("upgrade", indexMetaData);
+                        } else {
+                            generation = metaStateService.writeIndex("startup", indexMetaData);
                         }
+                        final long currentGeneration = generation;
+                        cleanupActions.add(() -> metaStateService.cleanupIndex(indexMetaData.getIndex(), currentGeneration));
+                        indices.put(indexMetaData.getIndex(), generation);
                     }
+
+                    final long metaStateGeneration =
+                            metaStateService.writeManifest("startup", new Manifest(globalStateGeneration, indices));
+                    cleanupActions.add(() -> metaStateService.cleanupMetaState(metaStateGeneration));
+                    performCleanup(cleanupActions);
```

Comment:
Ok, I've implemented what you've asked for. I've abandoned the use of `TwoPhaseCommit` interface because it's more suitable for non-related `TwoPhaseCommit` objects. Instead, I've added `Transaction` inner class in `GatewayMetaState`. The work is separated in 2 commits:
1) `MetaDataStateFormat` should perform new state files cleanup if `write` method call fails. https://github.com/elastic/elasticsearch/pull/35049/commits/fb3fc72fc84739a4572d42d77ecc27d8b57b1891
2) `GatewayMetaState` should perform new state files cleanup if any `writeXXX` method fails. https://github.com/elastic/elasticsearch/pull/35049/commits/20c9813bf637fe7a8c0b22e2078b098b3240ee4e